src/com/sap/piper/DefaultValueCache.groovy

```diff
@@ -29,4 +44,60 @@ class DefaultValueCache implements Serializable {
     static reset(){
         instance = null
     }
+
+
+    private static Set immutable(Set s) {
+        immutable(s, (Set)[])
+    }
+
+    private static List immutable(List l) {
+        immutable(l, (List)[])
+    }
+
+    private static def immutable(Collection _in, Collection _out) {
+
+        for(def e : _in) {
+            if(e in List || e in Set || e in Map) {
+                _out.add(immutable(e))
+            } else if (e in CharSequence && ! (e in String)) {
+                _out.add(e.toString())
+            } else {
+                typeCheck(e)
+                _out.add(e)
+            }
+        }
+        return _out.asImmutable()
+    }
+
+    private static Map immutable(Map m) {
+
+        Map result = [:]
+
+        for(def e : m.entrySet()) {
+
+            if(e.value in Map) {
+                result.put(e.key, immutable(e.value))
+            } else if (e.value in List || e.value in Set) {
+                result.put(e.key, immutable(e.value))
```

Comment:
Can be combined with the `Map` case, right?

Reply:
yes, done.